_posts/2019-09-30-introducing-arrow-flight.md

```diff
@@ -0,0 +1,260 @@
+remove the serialization costs associated with data transport and increase the
+overall efficiency of distributed data systems. Additionally, two systems that
+are already using Apache Arrow for other purposes can communicate data to each
+other with extreme efficiency.
+
+## Flight Basics
+
+The Arrow Flight libraries provide a development framework for implementing a
+service that can send and receive data streams. A Flight server supports
+several basic kinds of requests:
+
+* **ListFlights**: return a list of available data streams
+* **GetSchema**: return the schema for a data stream
+* **GetFlightInfo**: return a "query plan" for a dataset of interest, possibly
+  requiring consuming multiple data streams. This request can accept custom
+  serialized commands containing, for example, your specific application
+  parameters.
+* **DoGet**: send a data stream to a client
+* **DoPut**: receive a data stream from a client
+* **DoAction**: a perform an implementation-specific action and
+  return any results, i.e. a generalized function call
+* **ListActions**: return a list of available action types
+
+A simple Flight setup might consist of a single server to which clients connect
+and make DoGet requests.
+
+<div align="center">
+<img src="{{ site.baseurl }}/img/20191014_flight_simple.png"
+     alt="Flight Simple Architecture"
+     width="50%" class="img-responsive">
+</div>
+
+## Optimizing Data Throughput over gRPC
+
+While using a general-purpose messaging library like gRPC has numerous specific
+benefits beyond the obvious ones (taking advantage of all the engineering that
+Google has done on the problem).
+
+The best-supported way to use gRPC is to define services using an extended
+version of [Protocol Buffers][3] aka "Protobuf". A Protobuf plugin for gRPC
+generates a gRPC service that you can implement in your applications. RPC
+commands and data messages are serialized using the [Protobuf wire
+format][4]. Because we use "vanilla gRPC and Protocol Buffers", gRPC clients
+that are ignorant of the Arrow columnar format can still interact with Flight
+services and handle the Arrow data opaquely.
+
+The main data-related Protobuf type in Flight is called `FlightData`. Reading
+and writing Protobuf messages in general is not free, so we implemented some
+low-level optimizations in gRPC in both C++ and Java to do the following:
+
+* Generate the Protobuf wire format for `FlightData` including the Arrow record
+  batch being sent without going through any intermediate memory copying or
+  serialization steps.
+* Reconstruct a Arrow record batch from the Protobuf representation of
+  `FlightData` without any memory copying or deserialization.
+
+In a sense we are "having our cake and eat it, too". Flight implementations
+having these optimizations will have better performance, while naive gRPC
+clients talking to the Flight service and use a Protobuf library to deserialize
+`FlightData` (though with some performance penalty).
```

Comment:
Is a verb missing in this sentence?

Reply:
Fixed the grammar "clients can still talk to... and use a ..."